File: Graph/P1267CountServersThatCommunicate/CountServersThatCommunicate.java
package Graph.P1267CountServersThatCommunicate;

/**
 * @CLassName CountServersThatCommunicate
 * @Description TODO
 * @Author cst
 * @Date 1/30/21  3:19 PM
 */
public class CountServersThatCommunicate {
    private static CountServersThatCommunicate ONLY = null;
    public static CountServersThatCommunicate getInstance() {
        if (ONLY == null) {
            ONLY = new CountServersThatCommunicate();
        }
        return ONLY;
    }
    public int countServers(int[][] grid) {
        int res = 0;
        if (grid == null || grid[0].length == 0) {
            return res;
        }
        int m = grid.length;
        int n = grid[0].length;
        // 记录每一行/每一列拥有server的数量
        int[] row = new int[m];
        int[] col = new int[n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    row[i]++;
                    col[j]++;
                    res++;
                }
            }
        }

        // delete those who only has one server in the row/ column
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    if (row[i] == 1 && col[j] == 1){
                        res--;
                    }
                }
            }
        }
        return res;
    }
}

File: Graph/P1267CountServersThatCommunicate/CountServersThatCommunicate.py
import unittest
from typing import List

class Solution:
    def countServers(self, grid: List[List[int]]) -> int:
        res = 0
        if not grid:
            return res
        m, n = len(grid), len(grid[0])
        row, col = [0] * m, [0] * n
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    row[i] += 1
                    col[j] += 1
                    res += 1
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    if row[i] == 1 and col[j] == 1:
                        res -= 1
        return res            
        

class UnitTest(unittest.TestCase):
    def test(self):
        grid = [[1,0],[0,1]]
        self.assertEqual(0, Solution().countServers(grid))
        
        grid = [[1,1,0,0],[0,0,1,0],[0,0,1,0],[0,0,0,1]]
        self.assertEqual(4, Solution().countServers(grid))
        
if __name__ == "__main__":
    unittest.main()